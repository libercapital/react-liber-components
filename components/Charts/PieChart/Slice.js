"use strict";

require("core-js/modules/es6.object.define-property");

require("core-js/modules/es6.string.iterator");

require("core-js/modules/es6.weak-map");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

require("core-js/modules/es6.object.assign");

require("core-js/modules/es6.symbol");

require("core-js/modules/es6.array.index-of");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.iterator");

require("core-js/modules/es6.object.to-string");

require("core-js/modules/es6.object.keys");

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _d3Shape = require("d3-shape");

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Slice = function Slice(_ref) {
  var innerRadius = _ref.innerRadius,
      radius = _ref.radius,
      events = _ref.events,
      abs = _ref.abs,
      props = _objectWithoutProperties(_ref, ["innerRadius", "radius", "events", "abs"]);

  var pathFunctionBg = (0, _d3Shape.arc)().outerRadius(2 * abs + radius).innerRadius(innerRadius);
  var pathFunctionFg = (0, _d3Shape.arc)().outerRadius(abs / 2 + radius).innerRadius(innerRadius - abs / 2);
  var defaultPathFunction = (0, _d3Shape.arc)().outerRadius(radius).innerRadius(innerRadius);
  return _react["default"].createElement(_react.Fragment, null, _react["default"].createElement("g", events, _react["default"].createElement(_styles.Slice, _extends({}, props, {
    style: {
      fill: 'transparent'
    },
    pathFunction: pathFunctionBg
  })), _react["default"].createElement(_styles.Slice, _extends({}, props, {
    pathFunction: abs > 0 ? pathFunctionFg : defaultPathFunction
  }))));
};

Slice.propTypes = {
  slice: _propTypes["default"].objectOf(_propTypes["default"].any),
  // eslint-disable-line react/require-default-props
  style: _propTypes["default"].objectOf(_propTypes["default"].any),
  // eslint-disable-line react/require-default-props
  events: _propTypes["default"].objectOf(_propTypes["default"].any),
  // eslint-disable-line react/require-default-props
  pathFunction: _propTypes["default"].func,
  // eslint-disable-line react/require-default-props
  abs: _propTypes["default"].number,
  innerRadius: _propTypes["default"].number,
  radius: _propTypes["default"].number
};
Slice.defaultProps = {
  abs: 0,
  innerRadius: 0,
  radius: 100
};
var _default = Slice;
exports["default"] = _default;