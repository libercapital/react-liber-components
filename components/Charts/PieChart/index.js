"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _PieChart["default"];
  }
});
Object.defineProperty(exports, "Slice", {
  enumerable: true,
  get: function get() {
    return _Slice["default"];
  }
});

var _PieChart = _interopRequireDefault(require("./PieChart"));

var _Slice = _interopRequireDefault(require("./Slice"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }