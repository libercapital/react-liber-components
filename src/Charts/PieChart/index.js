export { default } from './PieChart';
export { default as Slice } from './Slice';
