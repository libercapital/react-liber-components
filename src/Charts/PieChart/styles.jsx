import styled from 'styled-components';
import { Slice as SliceRaw, VictoryContainer } from 'victory';

export const Slice = styled(SliceRaw)`
  transition: all 300ms;
  cursor: pointer;
`;

export const PieBox = styled(VictoryContainer)`
  width: unset !important;
  height: unset !important;
  & > svg {
    width: ${props => (props.width ? `${props.width}px` : 'unset')} !important;
    height: ${props => (props.height ? `${props.height}px` : 'unset')} !important;
  }
`;
