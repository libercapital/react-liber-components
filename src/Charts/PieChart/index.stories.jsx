import React from 'react';
import PieChart from './PieChart';

export default {
  title: 'Components 1.0|Charts/Pie Chart',
  parameters: {
    jest: ['PieChart'],
  },
};

export const defaultStory = () => <PieChart />;

defaultStory.story = {
  name: 'default',
};
