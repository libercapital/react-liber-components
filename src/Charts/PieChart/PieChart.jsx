import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { VictoryPie, VictoryTooltip } from 'victory';
import { handleThemeFromObject } from '../../Themes';

import Slice from './Slice';
import { PieBox } from './styles';

class PieChart extends Component {
  constructor() {
    super();
    this.state = {
      target: null,
      externalEvent: undefined,
      highlighted: false,
      current: null,
    };
  }


  componentDidUpdate(oldProps){
    const { targetEvent, target: nextTarget } = this.props;
    const { target } = this.state;
    if(oldProps.target !== nextTarget){
      this.setExternalEvent(nextTarget, target, targetEvent);
    }
  }

  setExternalEvent = (nextTarget, target) => {
    const hoverIn = [
      {
        target: ['data'],
        eventKey: 'all',
        mutation: this.handleOpacityIn,
        callback: () => { this.setState({ externalEvent: undefined }); },
      },
      {
        target: ['data'],
        eventKey: nextTarget,
        mutation: this.targetEventIn,
        callback: () => { this.setState({ externalEvent: undefined }); },
      },
    ];
    const hoverOut = [
      {
        target: ['data'],
        eventKey: 'all',
        mutation: this.handleOpacityOut,
        callback: () => { this.setState({ target: null, externalEvent: undefined }); },
      },
      {
        target: ['data'],
        eventKey: target,
        mutation: this.targetEventOut,
        callback: () => { this.setState({ target: null, externalEvent: undefined }); },
      },
    ];
    if (nextTarget && !target) {
      this.setState({
        target: nextTarget,
        externalEvent: hoverIn,
      });
    } else if (nextTarget && target && nextTarget !== target) {
      this.setState({
        externalEvent: [
          {
            target: ['data'],
            eventKey: target,
            mutation: this.targetEventOut,
            callback: () => {
              this.setState({
                target: nextTarget,
                externalEvent: hoverIn,
              });
            },
          },
        ],
      });
    } else if (!nextTarget && target) {
      this.setState({
        externalEvent: hoverOut,
      });
    }
  }


  getPadding = padding => {
    const paddingVal = typeof padding === 'number' ? padding : 0;
    const paddingObj = typeof padding === 'object' ? padding : {};
    return {
      top: paddingObj.top || paddingVal,
      bottom: paddingObj.bottom || paddingVal,
      left: paddingObj.left || paddingVal,
      right: paddingObj.right || paddingVal,
    };
  };

  getRadius = (width, height, propsPadding) => {
    const padding = this.getPadding(propsPadding);
    return (
      Math.min(
        width - padding.left - padding.right,
        height - padding.top - padding.bottom,
      ) / 2
    );
  };

  getTranslationSize = (dis, startAngle, endAngle) => {
    const angle = (startAngle + endAngle) / 2;
    const translationX = parseInt((Math.sin(angle) * dis).toFixed(5), 10);
    const translationY = parseInt((-Math.cos(angle) * dis).toFixed(5), 10);

    return {
      translationX,
      translationY,
      abs: Math.abs(translationX) === 0 ? dis : Math.abs(translationX),
    };
  };

  targetEventIn = props => {
    const { origin, slice, style, data, index } = props;
    const { translationX, translationY, abs } = this.getTranslationSize(
      10,
      slice.startAngle,
      slice.endAngle,
    );
    this.setState({ highlighted: true, current: index.toString() });
    return {
      abs,
      style: {
        fill: style.fill,
        transform:
          data.filter(item => Boolean(item.y) && Boolean(item.x)).length > 1
            ? `translate(${origin.x + translationX}px, ${origin.y +
                translationY}px)`
            : `translate(${props.origin.x}px, ${props.origin.x}px)`,
      },
    };
  };

  targetEventOut = props => {
    this.setState({ highlighted: false, current: null });
    return {
      abs: 0,
      style: {
        fill: props.style.fill,
        transform: `translate(${props.origin.x}px, ${props.origin.x}px)`,
      },
    };
  };

  handleOpacityIn = props => ({
    style: {
      fill: props.style.fill,
      opacity: 0.5,
      transform: `translate(${props.origin.x}px, ${props.origin.x}px)`,
    },
  });

  handleOpacityOut = props => ({
    style: {
      fill: props.style.fill,
      opacity: 1,
      transform: `translate(${props.origin.x}px, ${props.origin.x}px)`,
    },
  });

  render() {
    const {
      data,
      colorScale,
      height,
      width,
      innerRadius,
      padding,
      labels,
      labelRadius,
      x,
      y,
      onClick,
      onMouseEnter,
      onMouseLeave,
      targetEvent,
      theme,
      ...props
    } = this.props;
    const eventInData = {
      target: 'data',
      mutation: this.targetEventIn,
    };
    const eventOutData = {
      target: 'data',
      mutation: this.targetEventOut,
    };
    const opacityIn = {
      target: 'data',
      eventKey: 'all',
      mutation: this.handleOpacityIn,
    };
    const opacityOut = {
      target: 'data',
      eventKey: 'all',
      mutation: this.handleOpacityOut,
    };
    let eventHandlers;
    switch (targetEvent) {
      case 'click':
        eventHandlers = {
          onMouseEnter: (...args) => {
            onMouseEnter(...args);
          },
          onMouseLeave: (...args) => {
            onMouseLeave(...args);
          },
          onClick: (...args) => {
            onClick(...args);
            if (!this.props.disableInternalEvents) {
              if (!this.state.highlighted && !this.state.current) {
                return [opacityIn, eventInData];
              } else if (
                this.state.highlighted &&
                this.state.current !== args[2]
              ) {
                return [opacityOut, eventOutData, opacityIn, eventInData];
              }
              return [opacityOut, eventOutData];
            }
            return undefined;
          },
        };
        break;
      case 'hover':
      default:
        eventHandlers = {
          onMouseEnter: (...args) => {
            onMouseEnter(...args);
            if (!this.props.disableInternalEvents) {
              return [opacityIn, eventInData];
            }
            return undefined;
          },
          onMouseLeave: (...args) => {
            onMouseLeave(...args);
            if (!this.props.disableInternalEvents) {
              return [opacityOut, eventOutData];
            }
            return undefined;
          },
          onClick: (...args) => {
            onClick(...args);
          },
        };
        break;
    }
    const events = [
      {
        target: 'data',
        eventHandlers,
      },
    ];
    const radius = this.getRadius(width, height, padding);
    const slice = <Slice innerRadius={innerRadius} radius={radius} />;
    return (
      <div>
        <PieBox width={width} height={height}>
          <VictoryPie
            {...props}
            externalEventMutations={this.state.externalEvent}
            standalone={false}
            dataComponent={slice}
            height={height}
            width={width}
            padding={this.getPadding(padding)}
            innerRadius={innerRadius}
            data={data}
            events={events}
            colorScale={handleThemeFromObject(
              theme,
              'pieChart.colorScale',
              ['#18d8ee', '#537e86', '#2de8bf', '#40b7c9'],
              colorScale,
            )}
            labels={labels}
            labelRadius={labelRadius}
            style={{
              data: {
                transform: `translate(${width / 2}px, ${height / 2}px)`,
              },
            }}
            labelComponent={
              <VictoryTooltip
                height={25}
                width={25}
                orientation="top"
                style={{
                  fill: handleThemeFromObject(
                    theme,
                    'pieChart.tooltip',
                    '#009dff',
                  ),
                }}
                flyoutStyle={{
                  fill: handleThemeFromObject(
                    theme,
                    'pieChart.flyout.fill',
                    '#fafafa',
                  ),
                  stroke: handleThemeFromObject(
                    theme,
                    'pieChart.flyout.stroke',
                    '#cecece',
                  ),
                }}
                cornerRadius={0}
              />
            }
            x={x}
            y={y}
          />
        </PieBox>
      </div>
    );
  }
}

// const setExternalEvent = (nextTarget, target) => {
//   const hoverIn = [
//     {
//       target: ['data'],
//       eventKey: 'all',
//       mutation: PieChart.handleOpacityIn,
//       callback: () => {
//         this.setState({ externalEvent: undefined });
//       },
//     },
//     {
//       target: ['data'],
//       eventKey: nextTarget,
//       mutation: PieChart.targetEventIn,
//       callback: () => {
//         this.setState({ externalEvent: undefined });
//       },
//     },
//   ];
//   const hoverOut = [
//     {
//       target: ['data'],
//       eventKey: 'all',
//       mutation: PieChart.handleOpacityOut,
//       callback: () => {
//         this.setState({ target: null, externalEvent: undefined });
//       },
//     },
//     {
//       target: ['data'],
//       eventKey: target,
//       mutation: PieChart.targetEventOut,
//       callback: () => {
//         this.setState({ target: null, externalEvent: undefined });
//       },
//     },
//   ];
//   if (nextTarget && !target) {
//     return {
//       target: nextTarget,
//       externalEvent: hoverIn,
//     };
//   }
//   if (nextTarget && target && nextTarget !== target) {
//     return {
//       externalEvent: [
//         {
//           target: ['data'],
//           eventKey: target,
//           mutation: this.targetEventOut,
//           callback: () => {
//             this.setState({
//               target: nextTarget,
//               externalEvent: hoverIn,
//             });
//           },
//         },
//       ],
//     };
//   }
//   if (!nextTarget && target) {
//     return {
//       externalEvent: hoverOut,
//     };
//   }

//   return {};
// };


PieChart.propTypes = {
  data: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.any)),
  colorScale: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  height: PropTypes.number,
  width: PropTypes.number,
  innerRadius: PropTypes.number,
  padding: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.objectOf(PropTypes.number),
  ]),
  labels: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.func,
  ]),
  labelRadius: PropTypes.number,
  targetEvent: PropTypes.string,
  target: PropTypes.string,
  theme: PropTypes.objectOf(PropTypes.any),
  x: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.func,
  ]),
  y: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.func,
  ]),
  onClick: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  disableInternalEvents: PropTypes.bool,
};

PieChart.defaultProps = {
  data: [
    { x: 'A', y: 25 },
    { x: 'B', y: 10 },
    { x: 'C', y: 15 },
    { x: 'D', y: 50 },
  ],
  colorScale: undefined,
  height: 250,
  width: 250,
  innerRadius: 0,
  padding: 25,
  labels: [],
  labelRadius: 80,
  target: null,
  targetEvent: 'hover',
  theme: undefined,
  disableInternalEvents: false,
  x: 'x',
  y: 'y',
  onClick: () => {},
  onMouseEnter: () => {},
  onMouseLeave: () => {},
};

export default PieChart;
