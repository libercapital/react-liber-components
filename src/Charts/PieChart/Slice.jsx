import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { arc } from 'd3-shape';
import { Slice as VictorySlice } from './styles';

const Slice = ({ innerRadius, radius, events, abs, ...props }) => {
  const pathFunctionBg = arc()
    .outerRadius(2 * abs + radius)
    .innerRadius(innerRadius);
  const pathFunctionFg = arc()
    .outerRadius(abs / 2 + radius)
    .innerRadius(innerRadius - abs / 2);
  const defaultPathFunction = arc()
    .outerRadius(radius)
    .innerRadius(innerRadius);
  return (
    <Fragment>
      <g {...events}>
        <VictorySlice
          {...props}
          style={{ fill: 'transparent' }}
          pathFunction={pathFunctionBg}
        />

        <VictorySlice
          {...props}
          pathFunction={abs > 0 ? pathFunctionFg : defaultPathFunction}
        />
      </g>
    </Fragment>
  );
};

Slice.propTypes = {
  slice: PropTypes.objectOf(PropTypes.any), // eslint-disable-line react/require-default-props
  style: PropTypes.objectOf(PropTypes.any), // eslint-disable-line react/require-default-props
  events: PropTypes.objectOf(PropTypes.any), // eslint-disable-line react/require-default-props
  pathFunction: PropTypes.func, // eslint-disable-line react/require-default-props
  abs: PropTypes.number,
  innerRadius: PropTypes.number,
  radius: PropTypes.number,
};
Slice.defaultProps = {
  abs: 0,
  innerRadius: 0,
  radius: 100,
};

export default Slice;
