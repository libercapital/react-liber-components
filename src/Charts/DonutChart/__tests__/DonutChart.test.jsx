import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import PieChart from '../DonutChart';

describe('Test for DonutChart', () => {
  configure({ adapter: new Adapter() });
  const props = {
    onClick: sinon.spy(),
    onMouseEnter: sinon.spy(),
    onMouseLeave: sinon.spy(),
  };
  it('should render correctly', () => {
    const wrapper = shallow(<PieChart {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render correctly with innerRadius', () => {
    props.innerRadius = 50;
    const wrapper = shallow(<PieChart {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
