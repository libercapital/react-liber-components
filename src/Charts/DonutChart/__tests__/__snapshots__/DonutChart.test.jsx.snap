// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for DonutChart should render correctly 1`] = `
<PieChart
  data={
    Array [
      Object {
        "x": "A",
        "y": 25,
      },
      Object {
        "x": "B",
        "y": 10,
      },
      Object {
        "x": "C",
        "y": 15,
      },
      Object {
        "x": "D",
        "y": 50,
      },
    ]
  }
  disableInternalEvents={false}
  height={250}
  innerRadius={50}
  labelRadius={80}
  labels={Array []}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  padding={25}
  target={null}
  targetEvent="hover"
  width={250}
  x="x"
  y="y"
/>
`;

exports[`Test for DonutChart should render correctly with innerRadius 1`] = `
<PieChart
  data={
    Array [
      Object {
        "x": "A",
        "y": 25,
      },
      Object {
        "x": "B",
        "y": 10,
      },
      Object {
        "x": "C",
        "y": 15,
      },
      Object {
        "x": "D",
        "y": 50,
      },
    ]
  }
  disableInternalEvents={false}
  height={250}
  innerRadius={50}
  labelRadius={80}
  labels={Array []}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  padding={25}
  target={null}
  targetEvent="hover"
  width={250}
  x="x"
  y="y"
/>
`;
